semantics/src/main/kotlin/com/strumenta/kolasu/semantics/indexing/IndexWriter.kt

```diff
@@ -0,0 +1,66 @@
+package com.strumenta.kolasu.semantics.indexing
+
+import com.strumenta.kolasu.utils.sortBySubclassesFirst
+import kotlin.reflect.KClass
+import kotlin.reflect.safeCast
+
+/**
+ * Type-safe builder to create index writer instances.
+ * @param configuration the index writer configuration
+ * @return the configured index writer instance
+ **/
+fun indexWriter(
+    index: Index = InMemoryIndex(),
+    configuration: IndexWriterConfigurator.() -> Unit
+) = mutableMapOf<KClass<*>, (Any) -> Any?>()
+    .also { rules -> IndexWriterConfigurator(rules).apply(configuration) }
+    .let { rules -> IndexWriter(index, rules) }
+
+/**
+ * Configurable component handling write operations over an index.
+ * @property index the underlying index
+ * @property rules the index writer rules to use
+ **/
+class IndexWriter(
+    private val index: Index = InMemoryIndex(),
+    private val rules: Map<KClass<*>, (Any) -> Any?> = emptyMap()
+) {
+    /**
+     * Execute the index writer rule for the given [input].
+     * @param input the index writer rule input
+     * @return the output of the corresponding rule
+     **/
+    fun write(input: Any): Any? {
```

Comment:
To me this seems a symptom that this is overly generic. If it was a class created to work with Nodes, by looking at the signature I could figure that out. If instead the signatures are about using `Any` then it makes not obvious what this class can be used for

Reply:
This component is not meant to work with nodes only, but also types, symbols and any other things that might be stored in the index. 

I think it would be enough to have some documentation about the various components to make users understand what can be used and for what.

This and other comments give me the feeling that the whole solution for indexing should be redesigned. Perhaps we could postpone this into a later PR with a different architecture.

Otherwise, I am not sure how to proceed in addressing these comments and would accept suggestions/contributions.